docs/encyclopedia/application-message-passing.mdx

```diff
@@ -220,22 +220,39 @@ Every time the Workflow wakes up--generally, it wakes up when it needs to--it wi
 
 This execution is on a single thread–while this means you don’t have to worry about parallelism, you do need to worry about concurrency if you have written Signal and Update handlers that can block. These can run interleaved with the main Workflow and with one another, resulting in potential race conditions. These methods should be made reentrant.
 
+#### Initializing the Workflow first {#workflow-initializers}
+
+Notice from the [diagram](#message-handler-concurrency) that messages can be processed before the first time your Workflow's main method runs.
+
+In practice, this happens in two main scenarios:
+
+- You are using [Signal-with-Start](#signal-with-start), in which case the message handler always runs first.
+- Your Worker is delayed, such as when it's backlogged.
+
+So, it's important to initialize your Workflow's state before messages come in, otherwise you will be reading uninitialized instance variables.
+
+In object-oriented languages, you should annotate your constructor as a Workflow Initializer. Initializers take the same arguments as your Workflow's main method so that you can fully set up your state.
+
+Note that you cannot make blocking calls from your constructor. For that case, if you have a Signal or Update handler, you can make it [wait](#waiting).
+
+In other languages, wait to register your handlers until after you've done any initialization that your message handlers might need.
```

Comment:
@dandavison I kinda made this up re: Typescript.  (Slightly more confidence it's accurate for golang)  Not sure what to put here, or maybe I can be vague and punt to the lang-specific docs for other languages?

Reply:
That's good for TS. It's very similar to Go -- handler closure are registered by a call within the main workflow routine. @antlai-temporal described it similarly to you here: https://github.com/temporalio/documentation/pull/3107